cirq/testing/circuit_compare_test.py

```diff
@@ -543,3 +554,38 @@ class NoProtocol:
     with pytest.raises(AssertionError, match='disagree'):
         cirq.testing.assert_has_consistent_qid_shape(InconsistentOp3())
     cirq.testing.assert_has_consistent_qid_shape(NoProtocol())
+
+
+def test_assert_apply_unitary_works_when_axes_transposed_failure():
+
+    class BadOp:
+
+        def _apply_unitary_(self, args: cirq.ApplyUnitaryArgs):
+            # Get a more convenient view of the data.
+            a, b = args.axes
+            rest = list(range(len(args.target_tensor.shape)))
+            rest.remove(a)
+            rest.remove(b)
+            size = args.target_tensor.size
+            view = args.target_tensor.transpose([a, b, *rest])
+            view = view.reshape((4, size // 4))  # Oops. Reshape might copy.
+
+            # Apply phase gradient.
+            view[1, ...] *= 1j
+            view[2, ...] *= -1
+            view[3, ...] *= -1j
+            return args.target_tensor
+
+        def _num_qubits_(self):
+            return 2
+
+    assert cirq.has_unitary(BadOp())
+    bad_op = BadOp()
+
+    # Appears to work.
+    np.testing.assert_allclose(cirq.unitary(bad_op), np.diag([1, 1j, -1, -1j]))
+    # But fails the more discerning test.
+    with pytest.raises(AssertionError,
+                       match='acted differently on out-of-order axes'):
+        for _ in range(100):  # Axis orders chosen at random. Brute force a hit.
```

Comment:
This test should be deterministic. As is, it risks accidental random test failures and adds a bit to the creep of tests taking longer than necessary.

AFAICT, it's just a matter of finding and setting the seed that triggers the assert failure.

(Also, it's worth using getstate() and setstate() around the logic running with a fixed seed.)

Reply:
A quick numerical check indicates that an individual run will fail about 90% of the time. That suggests the chance 100 sequential checks failing is around 1 in 10^100. To my tastes that might as well be certain. In fact it's very likely that there is no failing random seed and it really *is* certain.

This is far less brittle than picking a seed and running one check, because picking a seed exposes us to the risk of the internals of the testing method changing how they generate the randomness.

Reply:
Fair enough.